### DISCUSSION TITLE
Idea: consolidate commands into an "fzf" mode

### DISCUSSION THREAD

**Discussion by pikeas** [PatrickF1/fzf.fish#103] [Ideas]
Great plugin, especially loving the recent QOL improvements in v5!

Have you considered using a single keystroke to enter "fzf mode"? I use ctrl-f the most by far, so my brain has rewired to think of that keystroke as fzf-ified <tab>. On the comparatively rarer occasions I'm committing code or want to check environment variables, I find myself typing `git add|set|ssh <ctrl-f>` only to be surprised by a list of files via fd. It then takes a few tries to remember the additional keybindings.

Your plugin has done *too* good a job of integrating fzf into fish, perhaps you can take it even further and turn it into a seamless "do what I mean" contextualized fuzzy tab completion?

**Comment by PatrickF1** [comment]
Hi @pikeas, thanks for the compliment!
I'm not sure what you mean. Are you saying you want a keystroke that enters fzf mode where you can then execute other keybindings to do other commands"

**Comment by pikeas** [comment]
In the current implementation, ctrl-f isn't context-dependent, it always runs fzf on the list of files output by fd. Tab completion, on the other hand, is context-dependent (command completion, ssh host completion, filename completion, etc).

So my idea is to combine these - context-aware input a la tab completion, fuzzy filtering/preview/etc a la fzf. 

* `git add <ctrl-f>` -> current behavior from `__fzf_search_git_status`
* `set <ctrl-f>` -> current behavior from `__fzf_search_shell_variables`
* (other possible contexts, for example `ssh <ctrl-f>`
* (default context) `<ctrl-f>` -> current behavior from `__fzf_search_current_dir`

**Comment by PatrickF1** [comment]
Hi @pikeas, what a wonderful idea! I really appreciate your creative thinking and I wish I had the time to expand the scope of my plugin so I could implement it. 

Let me explain more concretely why this is very difficult and I simply don't have the bandwidth since this is just a hobby for me
- The number of possible contexts to account for are very numerous.
- Checking for all the different contexts will result in complex, convoluted logic.
- Many contexts will be ambiguous. e.g. `vim <ctrl-f>` could be either `__fzf_search_git_status` or `__fzf_search_current_dir`. Sure, I could just pick one arbitrarily but from the user's perspective, they wouldn't be able to know and remember intuitively which command gets executed in those arbitrary contexts.
- Such a change is going to be backwards incompatible for everyone, which will be difficult to communicate on my end and difficult to adjust to for everyone else.
- Since the key bindings are mnemonic, I have hope that you'll soon develop the muscle memory to trigger the right one when you need it!

Hope you understand! I'm closing this issue but feel free to ask followup questions or send more ideas my way.

**Reply by autolyticus** [reply]
Quick question and reinterpreting what @pikeas said:

Does fish support command prefixes and key sequences as opposed to just key combinations? (A-la tmux)
Here's what I mean: Pressing Ctrl-F would put you into fzf mode where you have to enter a character which determines the action you take. 

So, pressing:
- Ctrl-F and then D could give me the directories in the current file, 
- Ctrl-F and then A could give me ALL the files (including hidden ones)

**Reply by PatrickF1** [reply]
> Does fish support command prefixes and key sequences as opposed to just key combinations? (A-la tmux)

Unfortunately no. that'd be super cool though and maybe worth bringing up with the fish team.

**Reply by pikeas** [reply]
Does [$fish_bind_mode](https://fishshell.com/docs/current/cmds/bind.html) help? `ctrl-f` -> set bind mode to "fzf" -> all other keybindings are configured as single letter. Minimal proof of concept:

```
bind \cF 'set fish_bind_mode fzf' # ctrl-F
bind -M fzf \e 'set -e fish_bind_mode' # esc
bind -M fzf g 'echo fzf git'
bind -M fzf f 'echo fzf files'
```

**Reply by PatrickF1** [reply]
Oh you're right, I can see that working. I'd just need to make sure to exit out of fzf mode after a command is done executing or if the user hits ESC. But, who knows what bugs might be triggered? To be honest, it seems like it'll be a lot of work and a HUGE disruptive change for all current users. I'll move this idea out of archive and reconsider it something in the future if I have more time.

**Reply by pikeas** [reply]
> Oh you're right, I can see that working. I'd just need to make sure to exit out of fzf mode after a command is done executing or if the user hits ESC. But, who knows what bugs might be triggered? To be honest, it seems like it'll be a lot of work and a HUGE disruptive change for all current users. I'll move this idea out of archive and reconsider it something in the future if I have more time.

Just kicked the tires on the following and it seems to work pretty well other than needing to tap an extra `esc`:

```
bind -M fzf d _fzf_search_directory
bind -M fzf l _fzf_search_git_log
bind -M fzf s _fzf_search_git_status
bind -M fzf r _fzf_search_history
bind -M fzf v '_fzf_search_variables (set --show | psub) (set --names | psub)'
bind \cF 'set fish_bind_mode fzf'
bind -M fzf \e 'set -e fish_bind_mode'
bind -M fzf \cC 'set -e fish_bind_mode'
# Binding '*' may also help, to clear the mode for all other keys
```

>  But, who knows what bugs might be triggered? To be honest, it seems like it'll be a lot of work and a HUGE disruptive change for all current users. I'll move this idea out of archive and reconsider it something in the future if I have more time.

Of course, no pressure! I love your plugin and respect your time - if you don't feel this change provides enough improvement to warrant the dev time, I understand. I'd do it myself and make a PR but I don't have a strong enough background in fish scripting and internals.

For the work required, is it possible that it's as simple as adding a small helper function that runs at the end of each command that unsets `$fish_bind_mode` if set?

**Reply by PatrickF1** [reply]
> if you don't feel this change provides enough improvement to warrant the dev time, I understand

Thanks for understanding. Honestly, I think this is a great idea and would be worth doing if this plugin hadn't already become so developed. The userbase of fzf.fish is already used very to using it in its current state. As a product engineer at work, I know how painful it is to cause customers to have to change their workflow. That's usually how you make people angry, and as a rule of thumb, you want minimize transitions/migrations as much as possible. I mean, just take a look at how much overhead went into 7.0 to make migration painless. 

So, the reality is that this should probably just be a separate plugin. Or I guess I can make the change and lose a bunch of users. Another reason to make it a separate plugin is now that the US is opening back up, I have a lot less free time than I used to and can't give this plugin the time it deserves. IMO, I'd love to see someone who has the time to create a better plugin.

> For the work required, is it possible that it's as simple as adding a small helper function that runs at the end of each command that unsets $fish_bind_mode if set?

Yes, I think that would work. Another edge case is if the user Ctrl+C out in the middle of the command. Then they'd be stuck. Not sure how to deal with that.

**Comment by nzig** [comment]
Please check out #214.
